marklogic-data-hub/src/main/java/com/marklogic/hub/step/impl/QueryStepRunner.java

```diff
@@ -0,0 +1,457 @@
+    }
+
+    @Override
+    public Job run(Collection uris) {
+        runningThread = null;
+        Job job = createJob();
+        return this.runHarmonizer(job,uris);
+    }
+
+    private Collection<String> runCollector() throws Exception {
+        Collector c = new CollectorImpl(this.flow, jobId);
+        c.setHubConfig(hubConfig);
+        c.setClient(stagingClient);
+
+        stepStatusListeners.forEach((StepStatusListener listener) -> {
+            listener.onStatusChange(this.jobId, 0, "running collector");
+        });
+
+        final DiskQueue<String> uris;
+        uris = c.run(this.flow.getName(), step, this.jobId, options);
+        return uris;
+    }
+
+    private Job runHarmonizer(Job job,Collection uris) {
+        AtomicLong successfulEvents = new AtomicLong(0);
+        AtomicLong failedEvents = new AtomicLong(0);
+        AtomicLong successfulBatches = new AtomicLong(0);
+        AtomicLong failedBatches = new AtomicLong(0);
+
+        stepStatusListeners.forEach((StepStatusListener listener) -> {
+            listener.onStatusChange(job.getJobId(), 0, "starting step execution");
+        });
+
+        if ( uris == null || uris.size() == 0 ) {
+            stepStatusListeners.forEach((StepStatusListener listener) -> {
+                listener.onStatusChange(job.getJobId(), 100, "collector returned 0 items");
+            });
+            stepFinishedListeners.forEach((StepFinishedListener::onStepFinished));
+            job.setCounts(0,0,0,0,0);
+            job.withStatus(JobStatus.COMPLETED_PREFIX + step);
+            jobUpdate.postJobs(jobId, JobStatus.COMPLETED_PREFIX + step, step);
+            return job;
+        }
+
+        Vector<String> errorMessages = new Vector<>();
+
+        dataMovementManager = stagingClient.newDataMovementManager();
+
+        double batchCount = Math.ceil((double) uris.size() / (double) batchSize);
+
+        HashMap<String, JobTicket> ticketWrapper = new HashMap<>();
+
+        ConcurrentHashMap<DatabaseClient, FlowResource> databaseClientMap = new ConcurrentHashMap<>();
+        Map<String,Object> fullResponse = new HashMap<>();
+        ObjectMapper mapper = new ObjectMapper();
+        queryBatcher = dataMovementManager.newQueryBatcher(uris.iterator())
+            .withBatchSize(batchSize)
+            .withThreadCount(threadCount)
+            .withJobId(job.getJobId())
+            .onUrisReady((QueryBatch batch) -> {
+                try {
+                    FlowResource flowResource;
+
+                    if (databaseClientMap.containsKey(batch.getClient())) {
+                        flowResource = databaseClientMap.get(batch.getClient());
+                    } else {
+                        flowResource = new FlowResource(batch.getClient(), destinationDatabase, flow);
+                        databaseClientMap.put(batch.getClient(), flowResource);
+                    }
+                    options.put("uris", batch.getItems());
```

Comment:
This is a shared Map Object so race conditions can occur when setting "uris". 

reqOptions
```suggestion
                    HashMap<> reqOptions = new HashMap<>(options);
                    reqOptions.put("uris", batch.getItems());
```